torch/csrc/distributed/autograd/engine/dist_engine.cpp

```diff
@@ -23,6 +23,52 @@ using torch::autograd::variable_list;
 static constexpr char* kNumBackwardPasses = "num_current_backward_passes";
 static constexpr char* kNumAutogradContexts = "num_autograd_contexts";
 
+// This hook does 3 things:
+//   1. Accumuate the gard to RPC context.
+//   2. Call pre hooks of the original AccumulateGrad to modify the input grad.
+//   3. Call post hooks of the original AccumulateGrad.
+struct DistAccumulateGradCaptureHook
+    : GraphTask::ExecInfo::Capture::GradCaptureHook {
+  DistAccumulateGradCaptureHook(
+      std::shared_ptr<AccumulateGrad> accumulateGrad,
+      ContextPtr autogradContext)
+      : accumulateGrad_(std::move(accumulateGrad)),
+        autogradContext_(std::move(autogradContext)) {}
+
+  torch::Tensor operator()(const torch::Tensor& grad) override {
+    variable_list inputGrads = {grad};
+    // It's intended that pre/post hooks are still called even if the grad is
+    // undenfined here.
+    for (const auto& hook : accumulateGrad_->pre_hooks()) {
+      inputGrads = (*hook)(inputGrads);
+      // A pre-hook should return a varaible list of the same size as the
+      // function's inputs.
+      TORCH_INTERNAL_ASSERT(inputGrads.size() == 1);
```

Comment:
nit: in the engine, such check does not exist (not sure why) so it's not necessary here. If you want to keep it (which makes sense here since this is an AccumulateGrad Node), I don't think it should be an internal assert though. This is user code that can return anything. I would lean towards something like: `TORCH_CHECK(inputGrads.size() == 1, "Pre hook for AccumulateGrad returned an invalid number of elements, got ", inputGrad.size(), " but 1 was expected");`.

Reply:
I'll remove those 2 checks.